src/test/validation_chainstatemanager_tests.cpp

```diff
@@ -431,12 +428,16 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_loadblockindex, TestChain100Setup)
     int num_indexes{0};
     int num_assumed_valid{0};
     const int expected_assumed_valid{20};
-    const int last_assumed_valid_idx{40};
+    const int last_assumed_valid_idx{111};
```

Comment:
In commit "test: adjust chainstate tests to use recognized snapshot base" (64f431029f04cc47889dbe8fad55ccc139af9c07)

One problem with this change is that it makes the test less meaningful, now the snapshot block is the tip of the chain, and there are no blocks after it, so the test can no longer check that blocks after the snapshot block are added to the snapshot chain's setBlockIndex set. But it seems like if you just mine 20 blocks on line 435 and check for 120 blocks on line 440, the test will pass with no other changes.

Also just in general I've been finding this test and the code around it very confusing, and think some extra comments and checks could help clarify it without having to change the existing code:

```diff
--- a/src/test/validation_chainstatemanager_tests.cpp
+++ b/src/test/validation_chainstatemanager_tests.cpp
@@ -422,17 +422,20 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_loadblockindex, TestChain100Setup)
 
     int num_indexes{0};
     int num_assumed_valid{0};
+    // Blocks in range [assumed_valid_start_idx, last_assumed_valid_idx) will be
+    // marked as assumed-valid and not having data.
     const int expected_assumed_valid{20};
     const int last_assumed_valid_idx{111};
     const int assumed_valid_start_idx = last_assumed_valid_idx - expected_assumed_valid;
 
-    // Get to a valid assumeutxo snapshot
-    mineBlocks(10);
+    // Mine to height 120, past the hardcoded regtest assumeutxo snapshot at
+    // height 110
+    mineBlocks(20);
 
     CBlockIndex* validated_tip{nullptr};
     CBlockIndex* assumed_base{nullptr};
     CBlockIndex* assumed_tip{WITH_LOCK(chainman.GetMutex(), return chainman.ActiveChain().Tip())};
-    BOOST_CHECK_EQUAL(assumed_tip->nHeight, 110);
+    BOOST_CHECK_EQUAL(assumed_tip->nHeight, 120);
 
     auto reload_all_block_indexes = [&]() {
         // For completeness, we also reset the block sequence counters to
@@ -458,7 +461,7 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_loadblockindex, TestChain100Setup)
         LOCK(::cs_main);
         auto index = cs1.m_chain[i];
 
-        // Blocks with heights in range [90, 110] are marked ASSUMED_VALID
+        // Blocks with heights in range [91, 110] are marked ASSUMED_VALID
         if (i < last_assumed_valid_idx && i >= assumed_valid_start_idx) {
             index->nStatus = BlockStatus::BLOCK_VALID_TREE | BlockStatus::BLOCK_ASSUMED_VALID;
         }
@@ -492,10 +495,36 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_loadblockindex, TestChain100Setup)
     // Set tip of the assume-valid-based chain to the assume-valid block
     cs2.m_chain.SetTip(*assumed_base);
 
+    // Sanity check test variables.
+    BOOST_CHECK_EQUAL(num_indexes, 121); // 121 total blocks, including genesis
+    BOOST_CHECK_EQUAL(assumed_tip->nHeight, 120);  // original chain has height 120
+    BOOST_CHECK_EQUAL(validated_tip->nHeight, 90); // current cs1 chain has height 90
+    BOOST_CHECK_EQUAL(assumed_base->nHeight, 110); // current cs2 chain has height 110
+
+    // Regenerate cs1.setBlockIndexCandidates and cs2.setBlockIndexCandidate and
+    // check contents below.
     reload_all_block_indexes();
 
-    // The fully validated chain should have the current validated tip
-    // and the assumed valid base as candidates.
+    // The fully validated chain should only have the current validated tip and
+    // the assumed valid base as candidates, blocks 90 and 110. Specifically:
+    //
+    // - It does not have blocks 0-89 because they contain less work than the
+    //   chain tip.
+    //
+    // - It has block 90 because it has data and equal work to the chain tip,
+    //   (since it is the chain tip).
+    //
+    // - It does not have blocks 91-109 because they do not contain data.
+    //
+    // - It has block 110 even though it does not have data, because
+    //   LoadBlockIndex has a special case to always add the snapshot block as a
+    //   candidate. The special case is only actually intended to apply to the
+    //   snapshot chainstate cs2, not the background chainstate cs1, but it is
+    //   written broadly and applies to both.
+    //
+    // - It does not have any blocks after height 110 because cs1 is a background
+    //   chainstate, and only blocks where are ancestors of the snapshot block
+    //   are added as candidates for the background chainstate.
     BOOST_CHECK_EQUAL(cs1.setBlockIndexCandidates.size(), 2);
     BOOST_CHECK_EQUAL(cs1.setBlockIndexCandidates.count(validated_tip), 1);
     BOOST_CHECK_EQUAL(cs1.setBlockIndexCandidates.count(assumed_base), 1);
@@ -503,8 +532,25 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_loadblockindex, TestChain100Setup)
     // The assumed-valid tolerant chain has the assumed valid base as a
     // candidate, but otherwise has none of the assumed-valid (which do not
     // HAVE_DATA) blocks as candidates.
+    //
+    // Specifically:
+    // - All blocks below height 110 are not candidates, because cs2 chain tip
+    //   has height 110 and they have less work than it does.
+    //
+    // - Block 110 is a candidate even though it does not have data, because it
+    //   is the snapshot block, which is assumed valid.
+    //
+    // - Blocks 111-120 are added because they have data.
+
+    // Check that block 90 is absent
     BOOST_CHECK_EQUAL(cs2.setBlockIndexCandidates.count(validated_tip), 0);
+    // Check that block 109 is absent
+    BOOST_CHECK_EQUAL(cs2.setBlockIndexCandidates.count(assumed_base->pprev), 0);
+    // Check that block 110 is present
+    BOOST_CHECK_EQUAL(cs2.setBlockIndexCandidates.count(assumed_base), 1);
+    // Check that block 120 is present
     BOOST_CHECK_EQUAL(cs2.setBlockIndexCandidates.count(assumed_tip), 1);
+    // Check that 11 blocks total are present.
     BOOST_CHECK_EQUAL(cs2.setBlockIndexCandidates.size(), num_indexes - last_assumed_valid_idx + 1);
 }
 
diff --git a/src/validation.cpp b/src/validation.cpp
index f77325daedb5..9c8815914a57 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -3462,7 +3462,8 @@ void Chainstate::ResetBlockFailureFlags(CBlockIndex *pindex) {
 void Chainstate::TryAddBlockIndexCandidate(CBlockIndex* pindex)
 {
     AssertLockHeld(cs_main);
-    // The block only is a candidate for the most-work-chain if it has more work than our current tip.
+    // The block only is a candidate for the most-work-chain if it has the same
+    // or more work than our current tip.
     if (m_chain.Tip() != nullptr && setBlockIndexCandidates.value_comp()(pindex, m_chain.Tip())) {
         return;
     }
```

I think it would be possible to make other improvements to the test beyond just adding comments. For example, I think it would be good to get rid of most of the variables and constants and just refer to blocks directly by height. Also, it would be good to add a fork to the chain below the snapshot height, to make sure blocks not leading up to the snapshot are not added as candidates to the background chainstate.